xds/src/main/java/io/grpc/xds/sds/ReferenceCountingSslContextProviderMap.java

```diff
@@ -0,0 +1,119 @@
+ *
+ * @param <K> Key type for the map
+ */
+@ThreadSafe
+final class ReferenceCountingSslContextProviderMap<K> {
+
+  private final HashMap<K, Instance> instances;
+  private final SslContextProviderFactory<K> sslContextProviderFactory;
+
+  ReferenceCountingSslContextProviderMap(SslContextProviderFactory<K> sslContextProviderFactory) {
+    checkNotNull(sslContextProviderFactory, "sslContextProviderFactory");
+    instances = new HashMap<>();
+    this.sslContextProviderFactory = sslContextProviderFactory;
+  }
+
+  /**
+   * Gets an existing instance of {@link SslContextProvider}. If it doesn't exist, creates a new one
+   * using the provided {@link SslContextProviderFactory&lt;K&gt;}
+   */
+  public SslContextProvider get(K key) {
+    checkNotNull(key, "key");
+    return getInternal(key);
+  }
+
+  /**
+   * Releases an instance of the given {@link SslContextProvider}.
+   *
+   * <p>The instance must have been obtained from {@link #get(K)}. Otherwise will throw
+   * IllegalArgumentException.
+   *
+   * <p>Caller must not release a reference more than once. It's advised that you clear the
+   * reference to the instance with the null returned by this method.
+   *
+   * @param key the key that identifies the shared resource
+   * @param value the instance to be released
+   * @return a null which the caller can use to clear the reference to that instance.
+   */
+  public SslContextProvider release(final K key, final SslContextProvider value) {
+    checkNotNull(key, "key");
+    checkNotNull(value, "value");
+    return releaseInternal(key, value);
+  }
+
+  private synchronized SslContextProvider getInternal(K key) {
+    Instance instance = instances.get(key);
+    if (instance == null) {
+      instance = new Instance(sslContextProviderFactory.createSslContextProvider(key));
+      instances.put(key, instance);
+    }
+    instance.refcount++;
+    return instance.sslContextProvider;
+  }
+
+  private synchronized SslContextProvider releaseInternal(
+      final K key, final SslContextProvider instance) {
+    final Instance cached = instances.get(key);
+    if (cached == null) {
+      throw new IllegalArgumentException("No cached instance found for " + key);
+    }
+    Preconditions.checkArgument(
+        instance == cached.sslContextProvider, "Releasing the wrong instance");
+    cached.refcount--;
+    if (cached.refcount == 0) {
+      try {
+        cached.sslContextProvider.close();
+      } finally {
+        instances.remove(key);
+      }
+    }
+    // Always return null
+    return null;
+  }
+
+  /** A factory to create an SslContextProvider from the given key. */
+  public interface SslContextProviderFactory<K> {
+    SslContextProvider createSslContextProvider(K key);
+  }
+
+  private static class Instance {
+    final SslContextProvider sslContextProvider;
+    int refcount;
+
+    Instance(SslContextProvider sslContextProvider) {
+      this.sslContextProvider = sslContextProvider;
+      this.refcount = 0;
```

Comment:
this should set to 1. we should treat the refcount==0 as impossible condition. also, `refCount` instead of `refcount`.

Reply:
PTAL and see if you still want to see refCount initialized to 1.

Reply:
object being created implies that it has refCount 1. if it is initially refCount = 0, it means it should be released. overall the layering (or separate of concern) is not clear (what the instant supposed to do is in the outside such as mutating refCount).

Reply:
couple of points:
- the refCount manipulation is now completely internal to the `Instance` class as per your suggestion so that part is addressed.
- the refCount logic is copied from `SharedResourceHolder` class and looked okay to me. Of course I can try to address your concern and see if I can initialize it to 1 by eliminating the increment in the first `get`.

Reply:
So refCount starts at 1 now.